storage/asynchbase/src/main/java/net/opentsdb/storage/Tsdb1xScanners.java

```diff
@@ -14,57 +14,965 @@
+    } else {
+      max_multi_get_cardinality = ((Tsdb1xHBaseDataStore) node.factory()).dynamicInt(Tsdb1xHBaseDataStore.MAX_MG_CARDINALITY_KEY);
+    }
+    
+    if (node.schema().rollupConfig() != null && rollup_usage != RollupUsage.ROLLUP_RAW) {
+      Downsampler ds = query.getMetrics().get(0).getDownsampler();
+      if (ds != null) {
+        rollup_intervals = node.schema().rollupConfig().getRollupInterval(
+            DateTime.parseDuration(ds.getInterval()) / 1000, ds.getInterval());
+      } else if (query.getTime().getDownsampler() != null) {
+        ds = query.getTime().getDownsampler();
+        rollup_intervals = node.schema().rollupConfig().getRollupInterval(
+            DateTime.parseDuration(ds.getInterval()) / 1000, ds.getInterval());
+      } else {
+        rollup_intervals = null;
+      }
+      
+      if (ds != null) {
+        rollup_aggregation = ds.getAggregator();
+      } else {
+        rollup_aggregation = null;
+      }
+    } else {
+      rollup_intervals = null;
+      rollup_aggregation = null;
+    }
+    
+    if (Strings.isNullOrEmpty(query.getMetrics().get(0).getAggregator()) ) {
+      rollup_group_by = query.getTime().getAggregator();
+    } else {
+      rollup_group_by = query.getMetrics().get(0).getAggregator();
+    }
+  }
+  
+  /**
+   * Call to fetch the next set of data from the scanners. 
+   * <b>WARNING:</b> Do not call this from the parent node without receiving
+   * a response first. Only one call can be outstanding at a time.
+   * 
+   * @param result A non-null result.
+   * @param span An optional span.
+   * @throws IllegalArgumentException if the result is null.
+   */
+  public void fetchNext(final Tsdb1xQueryResult result, final Span span) {
+    if (result == null) {
+      throw new IllegalArgumentException("Result must be initialized");
+    }
+    
+    synchronized (this) {
+      if (current_result != null) {
+        throw new IllegalStateException("Query result must have been null "
+            + "to start another query!");
+      }
+      current_result = result;
+    }
+    
+    // just extra safe locking. Shouldn't ever happen.
+    if (!initialized) {
+      synchronized (this) {
+        if (!initialized) {
+          initialize(span);
+          return;
+        } else {
+          throw new IllegalStateException("Lost initialization race. "
+              + "Who called me? This shouldn't happen");
+        }
+      }
+    }
+    
+    scanner_index = 0;
+    scanNext(span);
   }
   
-  boolean isFull() {
-    // TODO - implement
-    return false;
+  /** @return Whether or not a child scanner or config has thrown an 
+   * exception. */
+  boolean hasException() {
+    return has_failed;
   }
   
+  /** Called by a child when the scanner has finished it's current run. */
   void scannerDone() {
-    // TODO - implement
+    boolean send_upstream = false;
+    synchronized (this) {
+      scanners_done++;
```

Comment:
Likewise, is the value of this variable meant to be memory-consistent across threads?